### PR TITLE
Add vector mapping behavior support

### PR DESCRIPTION
**Pull request by ButternCream**
This adds support for creating a mapping for enums to a float array.
We're essentially creating a `map<enum, repeated float>` and we want to convert each of these entries to their own float array columns.

Take for example
```
enum Model {
    MODEL_UNSPECIFIED = 0;
    MODEL_GPT_3_5_XXX = 1;
    MODEL_GPT_4_0_XXX = 2;
  }

message ModelEmbedding {
  option (pgdb.v1.msg).nested_only = true;
  Model model = 1;
  repeated float embedding = 2 [(pgdb.v1.options) = {vector_size: 3}];
}

repeated ModelEmbedding model_embeddings = 8 [(pgdb.v1.options) = {message_behavoir: MESSAGE_BEHAVOIR_VECTOR}];
```

This would add 2 float array columns to the schema because of our 2 enums.
so we would get something like
`embedding_model_1` 
`embedding_model_2`

which correspond to their respective model enums (1 and 2)

### CODE REVIEW

File: example/models/food/v1/food.pb.pgdb.go [hunk 1]
```diff
@@ -1953,6 +2013,32 @@ func (x *PastaIngredientPBDataQueryType) Identifier() exp.IdentifierExpression {
...
+func (x *PastaIngredientmodel_embeddings_MODEL_GPT_3_5_XXXQueryType) Identifier() exp.IdentifierExpression {
+	return exp.NewIdentifierExpression("", x.tableName, x.column)
+}
+
+type PastaIngredientmodel_embeddings_MODEL_GPT_4_0_XXXQueryType struct {
+	column    string
+	tableName string
+}
+
+func (x *PastaIngredientDBQueryUnsafe) model_embeddings_MODEL_GPT_4_0_XXX() *PastaIngredientmodel_embeddings_MODEL_GPT_4_0_XXXQueryType {
```
Comment (pquerna) [inline example/models/food/v1/food.pb.pgdb.go:2034]: I think the name for this prob needs to be ctx.Name()?  becaue this function isn't exported right now?

Comment (pquerna) [reply example/models/food/v1/food.pb.pgdb.go:2034]: Also we should strings.TrimPrefix("MODEL_"* from the enum name i think?

File: internal/pgdb/pgdb_field.go [hunk 2]
```diff
@@ -373,5 +375,103 @@ func getCommonFields(ctx pgsgo.Context, m pgs.Message, ix *importTracker) ([]*fi
...
+			tempCtx := &fieldContext{
+				ExcludeNested: true,
+				IsVirtual:     true,
+				DB: &pgdb_v1.Column{
+					Name:               fmt.Sprintf("%s_%d", pgColName, enumValue.Value()),
+					Type:               "vector",
+					Nullable:           true,
+					OverrideExpression: fmt.Sprintf("vector(%d)", vectorSize),
+				},
+				GoName:   goNameString, // Generated go struct name
```
Comment (pquerna) [inline internal/pgdb/pgdb_field.go:458]: I think this needs to be:
```
ctx.Name(field).String() + "_" + trimEnumPrefix(ctx.Name(enumValue).String())

```
 where trim enum prefix would remove the common prefix string from the enum?

File: example/models/food/v1/food.pb.pgdb.go (hunk 1 above)
Comment (ButternCream) [reply example/models/food/v1/food.pb.pgdb.go]: added this in https://github.com/ductone/protoc-gen-pgdb/pull/66/commits/fb6ee23cf16d8fef81bb4acfbe1aa59e70da9030

File: internal/pgdb/pgdb_field.go (hunk 2 above)
Comment (ButternCream) [reply internal/pgdb/pgdb_field.go:459]: added this in https://github.com/ductone/protoc-gen-pgdb/pull/66/commits/fb6ee23cf16d8fef81bb4acfbe1aa59e70da9030

File: internal/pgdb/pgdb_field.go [hunk 3]
```diff
@@ -373,5 +376,103 @@ func getCommonFields(ctx pgsgo.Context, m pgs.Message, ix *importTracker) ([]*fi
...
+		}
+
+		// enum values
+		for _, enumValue := range enumField.Type().Enum().Values() {
+			if enumValue.Value() == 0 {
+				// skip the zero value
+				continue
+			}
+
+			goNameString := ctx.Name(field).String() + "_" + strings.TrimPrefix(ctx.Name(enumValue).String(), "MODEL_")
```
Comment (pquerna) [inline internal/pgdb/pgdb_field.go:447]: The TrimPrefix here should be based on the `enumField` name

File: internal/pgdb/pgdb_field.go [hunk 4]
```diff
@@ -373,5 +376,103 @@ func getCommonFields(ctx pgsgo.Context, m pgs.Message, ix *importTracker) ([]*fi
...
+
+			// TODO fix nullibity
```
Comment (pquerna) [inline internal/pgdb/pgdb_field.go:449]: old TODO

File: internal/pgdb/pgdb_index.go [hunk 5]
```diff
@@ -175,5 +175,94 @@ func getCommonIndexes(ctx pgsgo.Context, m pgs.Message) ([]*indexContext, error)
...
+	for _, field := range m.Fields() {
+		ext := pgdb_v1.FieldOptions{}
+		_, err := field.Extension(pgdb_v1.E_Options, &ext)
+		if err != nil {
+			return nil, fmt.Errorf("pgdb: getField: failed to extract Message extension from '%s': %w", field.FullyQualifiedName(), err)
+		}
+		if ext.MessageBehavoir != pgdb_v1.FieldOptions_MESSAGE_BEHAVOIR_VECTOR {
+			continue
+		}
+		if !field.Type().IsRepeated() {
```
Comment (pquerna) [inline internal/pgdb/pgdb_index.go:192]: this can be a shared function `assertFieldIsVectorShape()`?

File: internal/pgdb/pgdb_field.go (hunk 3 above)
Comment (ButternCream) [reply internal/pgdb/pgdb_field.go:447]: enumField name is just "Model"

Comment (pquerna) [reply internal/pgdb/pgdb_field.go:447]: Yeah, so, we should Trim the `strings.TrimPrefix(strings.ToUpper(enumField.SOMETHINGHERE) + "_")`?  not sure whats on the enumField to grab here

File: internal/pgdb/pgdb_index.go (hunk 5 above)
Comment (ButternCream) [reply internal/pgdb/pgdb_index.go]: https://github.com/ductone/protoc-gen-pgdb/pull/66/commits/d311d794661bff07d3d381d7843587d3e2a70fb2

File: internal/pgdb/pgdb_field.go [hunk 6]
```diff
@@ -389,46 +389,10 @@ func getCommonFields(ctx pgsgo.Context, m pgs.Message, ix *importTracker) ([]*fi
...
-				if err != nil {
-					return nil, fmt.Errorf("pgdb: getField: failed to extract Message extension from '%s': %w", field.FullyQualifiedName(), err)
-				}
-				if subExt.VectorSize == 0 {
-					panic(fmt.Errorf("pgdb: vector message must have vector_size set on repeated float field: %s", field.FullyQualifiedName()))
-				}
-				vectorSize = subExt.VectorSize
-				floatField = subField
-			}
+		enumField, floatField, vectorSize, err := pgdb_v1.GetFieldVectorShape(field)
```
Comment (pquerna) [inline internal/pgdb/pgdb_field.go:393]: this should be in the internal package, not in the pgpdb/v1 package -- that package is a runtime package, internal is code generation time

File: internal/pgdb/pgdb_field.go (hunk 3 above)
Comment (ButternCream) [reply internal/pgdb/pgdb_field.go:411]: what about `ModelEmbeddings_PastaIngredient_MODEL_GPT_3_5_XXX` -> `ModelEmbeddings_MODEL_GPT_3_5_XXX` ?

File: internal/pgdb/pgdb_field.go (hunk 6 above)
Comment (ButternCream) [reply internal/pgdb/pgdb_field.go:393]: ok I originally put it there then changed my mind 😓

Comment (ButternCream) [reply internal/pgdb/pgdb_field.go]: https://github.com/ductone/protoc-gen-pgdb/pull/66/commits/ee4ecbcba6b2f2d3c6f08f594fa411498d7dd6d5

File: example/models/food/v1/food.pb.pgdb.go [hunk 7]
```diff
@@ -1692,6 +1786,32 @@ func (x *PastaIngredientDBQueryBuilder) FTSData() *PastaIngredientFTSDataSafeOpe
...
+
+func (x *PastaIngredientModelEmbeddings_MODEL_GPT_3_5_XXXSafeOperators) Identifier() exp.IdentifierExpression {
+	return exp.NewIdentifierExpression("", x.tableName, x.column)
+}
+
+func (x *PastaIngredientDBQueryBuilder) ModelEmbeddings_MODEL_GPT_3_5_XXX() *PastaIngredientModelEmbeddings_MODEL_GPT_3_5_XXXSafeOperators {
+	return &PastaIngredientModelEmbeddings_MODEL_GPT_3_5_XXXSafeOperators{tableName: x.tableName, column: "pb$" + "model_embeddings_1"}
+}
+
+type PastaIngredientModelEmbeddings_MODEL_GPT_4_0_XXXSafeOperators struct {
```
Comment (pquerna) [inline example/models/food/v1/food.pb.pgdb.go:1802]: @ButternCream doesnt seem like the TrimPrefix is working as we think?

Comment (ButternCream) [reply example/models/food/v1/food.pb.pgdb.go]: https://github.com/ductone/protoc-gen-pgdb/pull/66/commits/abc3f6dd4671b5ff85eea12658f43c9faca52f53

**Review by pquerna — APPROVED** [review]

**Pull request merged**